ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -316,136 +299,104 @@ public ByteBuf compressChunk(ByteBuf chunkBuffer, boolean isFullChunk) {
+                  + compressedBuffer.readableBytes(), ex);
+        }
+        throw (CompressionException) rethrowException;
       }
-      throw (CompressionException) rethrowException;
+    } finally {
+      newCompressedBuffer.release();
     }
 
     // Decompress succeeded, emit metrics.
     compressionMetrics.decompressSuccessRate.mark();
-    compressionMetrics.decompressExpandSizeBytes.inc(decompressedBuffer.length - dataSize);
+    compressionMetrics.decompressExpandSizeBytes.inc(decompressedBuffer.remaining() - compressedBuffer.readableBytes());
     return Unpooled.wrappedBuffer(decompressedBuffer);
   }
 
   /**
-   * Convert ByteBuf to byte array with offset and size.
-   * It does not advance the byteBuf position.
+   * Convert a ByteBuf to a single buffer.  If source buffer is already single buffer, just return it.
+   * If it is consisted of multiple buffers, combine them into 1 direct buffer.
    *
    * @param byteBuf The buffer to read from.
-   * @return The triple of buffer, buffer offset, and data size.
+   * @return The new buffer.  The new buffer must be released when done reading.
    */
-  Triple<byte[], Integer, Integer> convertByteBufToByteArray(ByteBuf byteBuf) {
-
-    // Convert chunk's ByteBuf to byte[] before applying compression.
-    byte[] sourceBuffer;
-    int sourceBufferOffset;
-    int sourceDataSize = byteBuf.readableBytes();
-
-    // If ByteBuf is backed by an array.  Just access array directly.
-    if (byteBuf.hasArray()) {
-      sourceBuffer = byteBuf.array();
-      sourceBufferOffset = byteBuf.arrayOffset();
-      return Triple.of(sourceBuffer, sourceBufferOffset, sourceDataSize);
-    }
-
-    // If ByteBuf contains only 1 NIO buffer, unwrap the buffer.
+  private ByteBuf combineBuffer(ByteBuf byteBuf) {
+    // If the source buffer is already has 1 buffer, create a duplicate.
     if (byteBuf.nioBufferCount() == 1) {
-      ByteBuffer nioBuffer = byteBuf.nioBuffers()[0];
-      if (nioBuffer.hasArray()) {
-        sourceBuffer = nioBuffer.array();
-        sourceBufferOffset = nioBuffer.arrayOffset();
-        return Triple.of(sourceBuffer, sourceBufferOffset, sourceDataSize);
-      }
-    }
-
-    // Mark the chunk buffer.  We will restore the readerIndex if buffer is incompressible.
-    // ByteBuf is backed by other format. Regardless, copy the data into an array.
-    byteBuf.markReaderIndex();
-    try {
-      sourceBufferOffset = 0;
-      sourceBuffer = new byte[sourceDataSize];
-      byteBuf.readBytes(sourceBuffer);
-    }
-    finally {
-      // Assume no compression, revert the reader index (that advanced due to reading bytes).
-      byteBuf.resetReaderIndex();
+      return byteBuf.retainedDuplicate();
     }
 
-    return Triple.of(sourceBuffer, sourceBufferOffset, sourceDataSize);
+    // Copy the buffer out without changing index.
+    ByteBuffer newBuffer = ByteBuffer.allocateDirect(byteBuf.readableBytes());
+    byteBuf.getBytes(0, newBuffer);
```

Comment:
getBytes is probably better method to use here since it doesn't change the readerIndex. But i don't think the starting index should be 0. It should probably be byteBuf.readerIndex().